providers/implementations/kem/hybrid_kem.c

```diff
@@ -0,0 +1,260 @@
+                           "%s", ctx->info->alg[i].name);
+            return 0;
+        }
+    }
+    return ossl_hybrid_set_ctx_params(ctx, params);
+}
+
+int ossl_hybrid_kem_encapsulate_init(void *vctx, void *vauth,
+                                     const OSSL_PARAM params[])
+{
+    HYBRID_PKEY_CTX *ctx = (HYBRID_PKEY_CTX *)vctx;
+    HYBRID_PKEY *auth = (HYBRID_PKEY *)vauth;
+
+    return hybrid_kem_init(ctx, auth, params, &EVP_PKEY_encapsulate_init);
+}
+
+int ossl_hybrid_kem_decapsulate_init(void *vctx, void *vauth,
+                                     const OSSL_PARAM params[])
+{
+    HYBRID_PKEY_CTX *ctx = (HYBRID_PKEY_CTX *)vctx;
+    HYBRID_PKEY *auth = (HYBRID_PKEY *)vauth;
+
+    return hybrid_kem_init(ctx, auth, params, &EVP_PKEY_decapsulate_init);
+}
+
+static int hybrid_kem_auth_init(HYBRID_PKEY_CTX *ctx,
+                                HYBRID_PKEY *key, HYBRID_PKEY *auth,
+                                const OSSL_PARAM params[],
+                                int (*initf)(EVP_PKEY_CTX *ctx,
+                                             EVP_PKEY *auth,
+                                             const OSSL_PARAM params[]))
+{
+    unsigned int i;
+
+    if (!ossl_prov_is_running())
+        return 0;
+
+    for (i = 0; i < ctx->info->num_algs; i++) {
+        EVP_PKEY_CTX_free(ctx->ctxs[i]);
+        ctx->ctxs[i] = EVP_PKEY_CTX_new_from_pkey(ctx->libctx,
+                                                  key->keys[i], key->propq);
+        if (ctx->ctxs[i] == NULL || !(*initf)(ctx->ctxs[i], auth->keys[i], params))
+            return 0;
+    }
+    return ossl_hybrid_set_ctx_params(ctx, params);
+}
+
+int ossl_hybrid_kem_auth_encapsulate_init(void *vctx, void *vkey,
+                                          void *vauthpriv,
+                                          const OSSL_PARAM params[])
+{
+    HYBRID_PKEY_CTX *ctx = (HYBRID_PKEY_CTX *)vctx;
+    HYBRID_PKEY *key = (HYBRID_PKEY *)vkey;
+    HYBRID_PKEY *authpriv = (HYBRID_PKEY *)vauthpriv;
+
+    return hybrid_kem_auth_init(ctx, key, authpriv, params,
+                                &EVP_PKEY_auth_encapsulate_init);
+}
+
+int ossl_hybrid_kem_auth_decapsulate_init(void *vctx, void *vkey,
+                                          void *vauthpub,
+                                          const OSSL_PARAM params[])
+{
+    HYBRID_PKEY_CTX *ctx = (HYBRID_PKEY_CTX *)vctx;
+    HYBRID_PKEY *key = (HYBRID_PKEY *)vkey;
+    HYBRID_PKEY *authpub = (HYBRID_PKEY *)vauthpub;
+
+    return hybrid_kem_auth_init(ctx, key, authpub, params,
+                                &EVP_PKEY_auth_decapsulate_init);
+}
+
+int ossl_hybrid_kem_encapsulate(void *vctx, unsigned char *enc, size_t *enclen,
+                                unsigned char *secret, size_t *secretlen)
+{
+    HYBRID_PKEY_CTX *ctx = (HYBRID_PKEY_CTX *)vctx;
+    size_t a, b;
+    unsigned int i;
+
+    if (ctx->key_length == 0)
+        return 0;
+
+    if (enc == NULL) {
```

Comment:
prudent to also check secret == NULL?

Reply:
I don't think this would be beneficial but I've added additional NULL checks above.